lib/Sema/CSSimplify.cpp

```diff
@@ -10634,15 +10663,106 @@ ConstraintSystem::simplifyApplicableFnConstraint(
+        subKind, argumentsLoc, trailingClosureMatching);
 
     switch (matchCallResult) {
-    case SolutionKind::Error:
+    case SolutionKind::Error: {
+      auto resultTy = func2->getResult();
+
+      // If this is a call that constructs a callable type with
+      // trailing closure(s), closure(s) might not belong to
+      // the constructor but rather to implicit `callAsFunction`,
+      // there is no way to determine that without trying.
+      if (resultTy->isCallableNominalType(DC) &&
+          argumentList->hasAnyTrailingClosures()) {
+        auto *calleeLoc = getCalleeLocator(argumentsLoc);
+
+        bool isInit = false;
+        if (auto overload = findSelectedOverloadFor(calleeLoc)) {
+          isInit = bool(dyn_cast_or_null<ConstructorDecl>(
+              overload->choice.getDeclOrNull()));
+        }
+
+        if (!isInit)
+          return SolutionKind::Error;
+
+        auto &ctx = getASTContext();
+        auto numTrailing = argumentList->getNumTrailingClosures();
+
+        SmallVector<Argument, 4> newArguments;
+        SmallVector<Argument, 4> trailingClosures;
+
+        for (unsigned i = 0, n = argumentList->size(); i != n; ++i) {
+          if (argumentList->isTrailingClosureIndex(i)) {
+            trailingClosures.push_back(argumentList->get(i));
+          } else {
+            newArguments.push_back(argumentList->get(i));
+          }
+        }
+
+        // Original argument list with all the trailing closures removed.
+        auto *newArgumentList = ArgumentList::createParsed(
+            ctx, argumentList->getLParenLoc(), newArguments,
+            argumentList->getRParenLoc(),
+            /*firstTrailingClosureIndex=*/None);
+
+        auto trailingClosureTypes = func1->getParams().take_back(numTrailing);
+        // The original result type is going to become a result of
+        // implicit `.callAsFunction` instead since `.callAsFunction`
+        // is inserted between `.init` and trailing closures.
+        auto callAsFunctionResultTy = func1->getResult();
+
+        // The implicit replacement for original result type which
+        // represents a callable type produced by `.init` call.
+        auto callableType =
+            createTypeVariable(getConstraintLocator({}), /*flags=*/0);
+
+        // The original application type with all the trailing closures
+        // dropped from it and result replaced to the implicit variable.
+        func1 = FunctionType::get(func1->getParams().drop_back(numTrailing),
+                                  callableType, func1->getExtInfo());
+
+        auto matchCallResult = ::matchCallArguments(
+            *this, func2, newArgumentList, func1->getParams(),
+            func2->getParams(), subKind, argumentsLoc, trailingClosureMatching);
+
+        if (matchCallResult != SolutionKind::Solved)
+          return SolutionKind::Error;
+
+        // Note that `createImplicit` cannot be used here because it
```

Comment:
I see `ArgumentList::createParsed` do set implicit as `false`, so I wonder If it make sense in this case `setImplicit()` in `implicitCallArgumentList`?

Reply:
Yeah, I think it's worth it, thanks!

Reply:
Heh, there is no `setImplicit()` in argument list which is intentional, so I'm just going to add trailing index to `createImplicit` looks like there is no way around that.

Reply:
I see `ArgumentList::createParsed` just call `ArgumentList::create`(which is also public) that takes implicit as parameter, so maybe call `create` directly? :)

Reply:
Nah, `createImplicit` makes sense here, I just didn't want to add a parameter but that's fine too :)

Reply:
Ah awesome, thanks!

Reply:
Fixed!

Reply:
I think we forgot the comment :)

Reply:
Good catch, fixed!